core/src/test/scala/unit/kafka/server/AbstractFetcherThreadTest.scala

```diff
@@ -722,6 +728,105 @@ class AbstractFetcherThreadTest {
     }
   }
 
+  @Test
+  def testReplicaFetcherThreadHandlingPartitionFailure(): Unit = {
+
+    val partition1 = new TopicPartition("topic1", 0)
+    val partition2 = new TopicPartition("topic2", 0)
+
+    val fetcher = new MockFetcherThread {
+
+      override def processPartitionData(topicPartition: TopicPartition, fetchOffset: Long, partitionData: FetchData): Option[LogAppendInfo] = {
+        if (topicPartition == partition1) {
+          throw new KafkaException()
+        } else {
+          super.processPartitionData(topicPartition, fetchOffset, partitionData)
+        }
+      }
+
+    }
+
+    fetcher.setReplicaState(partition1, MockFetcherThread.PartitionState(leaderEpoch = 0))
+    fetcher.addPartitions(Map(partition1 -> offsetAndEpoch(0L, leaderEpoch = 0)))
+    fetcher.setLeaderState(partition1, MockFetcherThread.PartitionState(leaderEpoch = 0))
+
+    fetcher.setReplicaState(partition2, MockFetcherThread.PartitionState(leaderEpoch = 0))
+    fetcher.addPartitions(Map(partition2 -> offsetAndEpoch(0L, leaderEpoch = 0)))
+    fetcher.setLeaderState(partition2, MockFetcherThread.PartitionState(leaderEpoch = 0))
+
+    // processing data fails for partition1
+    fetcher.doWork()
+
+    // partition1 marked as failed
+    assertTrue(failedPartitions.contains(partition1))
+    assertTrue(fetcher.fetchState(partition1).isEmpty)
+
+    // fetcher continues with rest of the partitions
+    assertEquals(Some(Fetching), fetcher.fetchState(partition2).map(_.state))
+    assertFalse(failedPartitions.contains(partition2))
+
+    // simulate a leader epoch
+    fetcher.removePartitions(Set(partition1))
+    failedPartitions.remove(Set(partition1))
+    fetcher.addPartitions(Map(partition1 -> offsetAndEpoch(0L, leaderEpoch = 1)))
+
+    fetcher.doWork()
+
+    // partition1 added back
+    assertTrue(fetcher.fetchState(partition1).nonEmpty)
+    assertFalse(failedPartitions.contains(partition1))
+
+  }
+
+  @Test
+  def testTruncationHandlingPartitionFailure(): Unit = {
```

Comment:
The code in testTruncationHandlingPartitionFailure() is almost identical to testReplicaFetcherThreadHandlingPartitionFailure(). Perhaps we can share the code by passing in a different fetcher?

Reply:
Yeah, I modified the tests to reuse the code.